File: src/main/java/org/panda_lang/reposilite/repository/LookupService.java
/*
 * Copyright (c) 2020 Dzikoysk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.panda_lang.reposilite.repository;

import com.google.api.client.http.GenericUrl;
import com.google.api.client.http.HttpRequest;
import com.google.api.client.http.HttpRequestFactory;
import com.google.api.client.http.HttpResponse;
import com.google.api.client.http.javanet.NetHttpTransport;
import io.javalin.http.Context;
import org.apache.commons.io.IOUtils;
import org.panda_lang.reposilite.Reposilite;
import org.panda_lang.reposilite.auth.Authenticator;
import org.panda_lang.reposilite.auth.Session;
import org.panda_lang.reposilite.config.Configuration;
import org.panda_lang.reposilite.metadata.MetadataService;
import org.panda_lang.reposilite.metadata.MetadataUtils;
import org.panda_lang.reposilite.utils.FilesUtils;
import org.panda_lang.reposilite.utils.Result;
import org.panda_lang.utilities.commons.text.ContentJoiner;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Arrays;
import java.util.Objects;

public final class LookupService {

    private final Configuration configuration;
    private final Authenticator authenticator;
    private final MetadataService metadataService;
    private final RepositoryService repositoryService;
    private final HttpRequestFactory requestFactory;

    public LookupService(Reposilite reposilite) {
        this.configuration = reposilite.getConfiguration();
        this.authenticator = reposilite.getAuthenticator();
        this.metadataService = reposilite.getMetadataService();
        this.repositoryService = reposilite.getRepositoryService();
        this.requestFactory = configuration.getProxied().isEmpty() ? null : new NetHttpTransport().createRequestFactory();
    }

    Result<Context, String> serveProxied(Context context) {
        String uri = context.req.getRequestURI();

        for (String proxied : configuration.getProxied()) {
            try {
                HttpRequest remoteRequest = requestFactory.buildGetRequest(new GenericUrl(proxied + uri));
                remoteRequest.setThrowExceptionOnExecuteError(false);
                HttpResponse remoteResponse = remoteRequest.execute();

                if (!remoteResponse.isSuccessStatusCode()) {
                    continue;
                }

                return Result.ok(context.status(remoteResponse.getStatusCode())
                        .contentType(remoteResponse.getContentType())
                        .header("Content-Length", Objects.toString(remoteResponse.getHeaders().getContentLength(), "0"))
                        .result(remoteResponse.getContent()));
            } catch (IOException e) {
                Reposilite.getLogger().warn("Proxied repository " + proxied + " is unavailable: " + e.getMessage());
            }
        }

        return Result.error("Artifact not found in local and remote repository");
    }

    Result<Context, String> serveLocal(Context context) {
        if (configuration.isFullAuthEnabled()) {
            Result<Session, String> authResult = this.authenticator.authUri(context);

            if (authResult.getError().isDefined()) {
                return Result.error(authResult.getError().get());
            }
        }

        String[] path = RepositoryUtils.normalizeUri(configuration, context.req.getRequestURI()).split("/");

        // discard invalid requests (less than 'repository/group/artifact')
        if (path.length < 3) {
            return Result.error("Unsupported request");
        }

        // remove empty element '/x/y/z' before root due to String#split result
        if (path[0].isEmpty()) {
            path = Arrays.copyOfRange(path, 1, path.length);
        }

        Repository repository = repositoryService.getRepository(path[0]);

        if (repository == null) {
            return Result.error("Repository " + path[0] + " not found");
        }

        // remove repository name from path
        String[] requestPath = Arrays.copyOfRange(path, 1, path.length);

        // discard invalid requests (less than 'group/(artifact OR metadata)')
        if (requestPath.length < 2) {
            return Result.error("Missing artifact path");
        }

        String requestedFileName = requestPath[requestPath.length - 1];

        if (requestedFileName.equals("maven-metadata.xml")) {
            try {
                String result = metadataService.generateMetadata(repository, requestPath);

                if (result == null) {
                    return Result.error("Metadata not found");
                }

                return Result.ok(context.contentType("text/xml").result(result));
            } catch (IOException e) {
                return Result.error("Cannot find metadata file");
            }
        }

        // resolve snapshot requests
        if (requestedFileName.contains("-SNAPSHOT")) {
            // ignore inspection to make it clear that 'requestPath' content is updated
            // noinspection ConstantConditions
            requestPath = repositoryService.resolveSnapshot(repository, requestPath);
        }

        Artifact artifact = repository.get(requestPath);

        if (artifact == null) {
            return Result.error("Artifact " + ContentJoiner.on("/").join(requestPath) + " not found");
        }

        requestedFileName = requestPath[requestPath.length - 1];
        File file = artifact.getFile(requestedFileName);

        if (!file.exists()) {
            Reposilite.getLogger().warn("File " + file.getAbsolutePath() + " does not exist");
            return Result.error("Artifact " + file.getName() + " not found");
        }

        FileInputStream content = null;

        try {
            String mimeType = Files.probeContentType(file.toPath());
            context.res.setContentType(mimeType);

            content = new FileInputStream(file);
            IOUtils.copy(content, context.res.getOutputStream());

            context.res.setContentLength((int) file.length());
            context.res.setHeader("Content-Disposition", "attachment; filename=\"" + MetadataUtils.getLast(path) + "\"");

            Reposilite.getLogger().info("Available: " + content.available() + "; mime: " + mimeType + "; size: " + file.length() + "; file: " + file.getPath());
            return Result.ok(context.header("Content-Disposition", "attachment; filename=\"" + MetadataUtils.getLast(path) + "\""));
        } catch (FileNotFoundException e) {
            Reposilite.getLogger().warn("Cannot read file " + file.getAbsolutePath());
            return Result.error("Cannot read artifact");
        } catch (IOException e) {
            return Result.error("Unknown mime type " + file.getName());
        } finally {
            FilesUtils.close(content);
        }
    }

}

File: reposilite-backend/src/main/kotlin/org/panda_lang/reposilite/maven/LookupService.kt
/*
 * Copyright (c) 2021 dzikoysk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.panda_lang.reposilite.maven

import io.javalin.http.HttpCode.NOT_FOUND
import io.javalin.http.HttpCode.UNAUTHORIZED
import org.panda_lang.reposilite.failure.api.ErrorResponse
import org.panda_lang.reposilite.failure.api.errorResponse
import org.panda_lang.reposilite.maven.api.LookupRequest
import org.panda_lang.reposilite.maven.api.LookupResponse
import org.panda_lang.reposilite.maven.api.Repository
import org.panda_lang.reposilite.web.orNull
import org.panda_lang.utilities.commons.function.Result
import java.nio.file.Path

internal class LookupService(
    private val repositoryService: RepositoryService,
    private val repositorySecurityProvider: RepositorySecurityProvider
) {

    fun lookup(lookupRequest: LookupRequest): Result<LookupResponse, ErrorResponse> {
        val repository = repositoryService.getRepository(lookupRequest.repository) ?: return errorResponse(NOT_FOUND, "Repository not found")
        val gav = Path.of(lookupRequest.gav)

        if (repository.isDirectory(gav) && repositorySecurityProvider.canBrowseResource(lookupRequest.accessToken, repository, gav).not()) {
            return errorResponse(UNAUTHORIZED, "Unauthorized indexing request")
        }
        else if (repositorySecurityProvider.canAccessResource(lookupRequest.accessToken, repository, gav).not()) {
            return errorResponse(UNAUTHORIZED, "Unauthorized access request")
        }

        return repository.getFileDetails(gav)
            .map { LookupResponse(it, repository.getFile(lookupRequest.gav).orNull()) }
    }

    fun findAllRepositories(): Collection<Repository> =
        repositoryService.getRepositories()

    /*
    fun exists(context: ReposiliteContext): Boolean {
        val uri: String = context.uri
        val result = repositoryAuthenticator.authDefaultRepository(context.header, uri)

        if (result.isErr) {
            // Maven requests maven-metadata.xml file during deploy for snapshot releases without specifying credentials
            // https://github.com/dzikoysk/reposilite/issues/184
            return if (uri.contains("-SNAPSHOT") && uri.endsWith(METADATA_FILE)) {
                false
            } else false
        }

        val path = result.get().key

        // discard invalid requests (less than 'group/(artifact OR metadata)')
        if (path.nameCount < 2) {
            return false
        }

        val repository = result.get().value
        return repository.exists(path)
    }

    fun find(context: ReposiliteContext): Result<LookupResponse, ErrorResponse> {
        val uri: String = context.uri
        val result = repositoryAuthenticator.authDefaultRepository(context.header, uri)

        if (result.isErr) {
            // Maven requests maven-metadata.xml file during deploy for snapshot releases without specifying credentials
            // https://github.com/dzikoysk/reposilite/issues/184
            return if (uri.contains("-SNAPSHOT") && uri.endsWith(METADATA_FILE)) {
                ResponseUtils.error(HttpStatus.SC_NOT_FOUND, result.error.message)
            } else Result.error(result.error)
        }

        var path = result.get().key

        // discard invalid requests (less than 'group/(artifact OR metadata)')
        if (path!!.nameCount < 2) {
            return ResponseUtils.error(HttpStatus.SC_NON_AUTHORITATIVE_INFORMATION, "Missing artifact identifier")
        }

        val repository = result.get().value
        val requestedFileName = path.fileName.toString()

        if (requestedFileName == "maven-metadata.xml") {
            return repository.getFile(path).map { LookupResponse.of("text/xml", Arrays.toString(it)) }
        }

        // resolve requests for latest version of artifact
        if (requestedFileName.equals("latest", ignoreCase = true)) {
            val requestDirectory = path.parent
            val versions: Result<List<Path>, ErrorResponse> = toSortedVersions(repository, requestDirectory)

            if (versions.isErr) {
                return versions.map { null }
            }

            val version = versions.get().firstOrNull()
                ?: return ResponseUtils.error(HttpStatus.SC_NOT_FOUND, "Latest version not found")

            return Result.ok(LookupResponse.of("text/plain", version.fileName.toString()))
        }

        // resolve snapshot requests
        if (requestedFileName.contains("-SNAPSHOT")) {
            path = repositoryService.resolveSnapshot(repository, path)

            if (path == null) {
                return Result.error(ErrorResponse(HttpStatus.SC_NOT_FOUND, "Latest version not found"))
            }
        }

        if (repository.isDirectory(path)) {
            return ResponseUtils.error(HttpStatus.SC_NON_AUTHORITATIVE_INFORMATION, "Directory access")
        }

        val bytes = repository.getFile(path)

        if (bytes.isErr) {
            return bytes.map { null }
        }

        val fileDetailsResult = repository.getFileDetails(path)

        return if (fileDetailsResult.isOk) {
            val fileDetails = fileDetailsResult.get()

            if (context.method != "HEAD") {
                context.output { it.write(bytes.get()) }
            }

            context.logger.debug("RESOLVED $path; mime: ${fileDetails.contentType}; size: ${repository.getFileSize(path).get()}")
            Result.ok(LookupResponse.of(fileDetails))
        }
        else {
            Result.error(fileDetailsResult.error)
        }
    }

     */
}